### PR TITLE
Cache ClusterConfig

### PR DESCRIPTION
**Pull request by Kahbazi**
Fixes https://github.com/microsoft/reverse-proxy/issues/310

### CODE REVIEW

File: src/ReverseProxy/Middleware/AffinitizedDestinationLookupMiddleware.cs [hunk 1]
```diff
@@ -41,15 +41,17 @@ internal class AffinitizedDestinationLookupMiddleware
 
         public Task Invoke(HttpContext context)
         {
-            var cluster = context.GetRequiredCluster();
+            var clusterConfig = context.GetRequiredClusterConfig();
 
-            var options = cluster.Config.Value?.SessionAffinityOptions ?? default;
+            var options = clusterConfig?.SessionAffinityOptions ?? default;
```
Comment (Tratcher) [inline src/ReverseProxy/Middleware/AffinitizedDestinationLookupMiddleware.cs:46]:
```suggestion
            var options = clusterConfig.SessionAffinityOptions;
```

File: src/ReverseProxy/Middleware/AffinitizedDestinationLookupMiddleware.cs [hunk 2]
```diff
@@ -41,15 +41,17 @@ internal class AffinitizedDestinationLookupMiddleware
...
 
             if (!options.Enabled)
             {
                 return _next(context);
             }
 
+            var cluster = context.GetRequiredCluster();
+
             return InvokeInternal(context, options, cluster);
```
Comment (Tratcher) [inline src/ReverseProxy/Middleware/AffinitizedDestinationLookupMiddleware.cs:55]: InvokeInternal only needs the clusterId, so let's pass that directly.

File: src/ReverseProxy/Middleware/AffinitizeRequestMiddleware.cs [hunk 3]
```diff
@@ -37,8 +37,8 @@ internal class AffinitizeRequestMiddleware
 
         public Task Invoke(HttpContext context)
         {
-            var cluster = context.GetRequiredCluster();
-            var options = cluster.Config.Value?.SessionAffinityOptions ?? default;
+            var clusterConfig = context.GetRequiredClusterConfig();
+            var options = clusterConfig?.SessionAffinityOptions ?? default;
```
Comment (Tratcher) [inline src/ReverseProxy/Middleware/AffinitizeRequestMiddleware.cs:41]:
```suggestion
            var options = clusterConfig.SessionAffinityOptions;
```

File: src/ReverseProxy/Middleware/LoadBalancingMiddleware.cs [hunk 4]
```diff
@@ -35,11 +35,11 @@ internal class LoadBalancingMiddleware
...
         public Task Invoke(HttpContext context)
         { 
-            var cluster = context.GetRequiredCluster();
+            var clusterConfig = context.GetRequiredClusterConfig();
             var destinationsFeature = context.GetRequiredDestinationFeature();
             var destinations = destinationsFeature.Destinations;
 
-            var loadBalancingOptions = cluster.Config.Value?.LoadBalancingOptions
+            var loadBalancingOptions = clusterConfig?.LoadBalancingOptions
                 ?? new ClusterConfig.ClusterLoadBalancingOptions(default);
```
Comment (Tratcher) [inline src/ReverseProxy/Middleware/LoadBalancingMiddleware.cs:43]:
```suggestion
            var loadBalancingOptions = clusterConfig.LoadBalancingOptions;
```

File: src/ReverseProxy/Middleware/DestinationInitializerMiddleware.cs [hunk 5]
```diff
@@ -56,6 +56,7 @@ public Task Invoke(HttpContext context)
             }
 
             context.Features.Set(cluster);
+            context.Features.Set(cluster.Config.Value);
```
Comment (Tratcher) [inline src/ReverseProxy/Middleware/DestinationInitializerMiddleware.cs:59]: We require this to not be null when we retrieve it, but we don't check if it's null before we store it. Please add a null check like the one on line 35.

File: src/ReverseProxy/Middleware/IReverseProxyFeature.cs [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+using System.Collections.Generic;
+using Microsoft.ReverseProxy.Abstractions;
+using Microsoft.ReverseProxy.RuntimeModel;
+
+namespace Microsoft.ReverseProxy.Middleware
+{
+    /// <summary>
+    /// Store current Cluster, ClusterConfig and Tracks proxy cluster destinations that are available to handle the current request.
+    /// </summary>
+    public interface IReverseProxyFeature
```
Comment (Kahbazi) [inline src/ReverseProxy/Middleware/IReverseProxyFeature.cs:13]: Any suggestion for the name?

File: src/ReverseProxy/Middleware/AffinitizeRequestMiddleware.cs [hunk 7]
```diff
@@ -33,16 +33,17 @@ internal class AffinitizeRequestMiddleware
...
-            var cluster = context.GetRequiredCluster();
-            var options = cluster.Config.Value?.SessionAffinityOptions ?? default;
+            var reverseProxy = context.GetReverseProxyFeature();
+            var options = reverseProxy.ClusterConfig.SessionAffinityOptions;
 
             if (options.Enabled)
             {
-                var destinationsFeature = context.GetRequiredDestinationFeature();
-                var candidateDestinations = destinationsFeature.Destinations;
+                var destinationsFeature = context.GetReverseProxyFeature();
```
Comment (Tratcher) [inline src/ReverseProxy/Middleware/AffinitizeRequestMiddleware.cs:41]: Remove duplicate call to GetReverseProxyFeature.

File: src/ReverseProxy/Middleware/HttpContextFeaturesExtensions.cs [hunk 8]
```diff
@@ -14,14 +14,20 @@ public static ClusterInfo GetRequiredCluster(this HttpContext context)
             return context.Features.Get<ClusterInfo>() ?? throw new InvalidOperationException("Cluster unspecified.");
```
Comment (Tratcher) [inline src/ReverseProxy/Middleware/HttpContextFeaturesExtensions.cs:14]: GetRequiredCluster should call GetRequiredRouteConfig.

File: src/ReverseProxy/Middleware/DestinationInitializerMiddleware.cs [hunk 9]
```diff
@@ -56,7 +59,11 @@ public Task Invoke(HttpContext context)
             }
 
             context.Features.Set(cluster);
```
Comment (Tratcher) [inline src/ReverseProxy/Middleware/DestinationInitializerMiddleware.cs:61]: This is available from the RouteConfig.
```suggestion
```

File: src/ReverseProxy/Middleware/DestinationInitializerMiddleware.cs [hunk 10]
```diff
@@ -36,15 +36,18 @@ public Task Invoke(HttpContext context)
             if (cluster == null)
```
Comment (Tratcher) [inline src/ReverseProxy/Middleware/DestinationInitializerMiddleware.cs:36]: Call the new GetRequiredRouteConfig API above?

File: src/ReverseProxy/Middleware/HttpContextFeaturesExtensions.cs [hunk 11]
```diff
@@ -14,14 +14,20 @@ public static ClusterInfo GetRequiredCluster(this HttpContext context)
...
+            var endpoint = context.GetEndpoint()
+               ?? throw new InvalidOperationException($"Routing Endpoint wasn't set for the current request.");
+
+            var routeConfig = endpoint.Metadata.GetMetadata<RouteConfig>()
+                ?? throw new InvalidOperationException($"Routing Endpoint is missing {typeof(RouteConfig).FullName} metadata.");
+
+            return routeConfig;
+        }
+
+        public static IReverseProxyFeature GetReverseProxyFeature(this HttpContext context)
```
Comment (Tratcher) [inline src/ReverseProxy/Middleware/HttpContextFeaturesExtensions.cs:28]:
```suggestion
        public static IReverseProxyFeature GetRequiredProxyFeature(this HttpContext context)
```

File: src/ReverseProxy/Middleware/IReverseProxyFeature.cs (hunk 6 above)
Comment (Tratcher) [reply src/ReverseProxy/Middleware/IReverseProxyFeature.cs:13]: *shrug* this is good enough for now.

File: src/ReverseProxy/Middleware/IReverseProxyFeature.cs [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+// Licensed under the MIT License.
+
+using System.Collections.Generic;
+using Microsoft.ReverseProxy.Abstractions;
+using Microsoft.ReverseProxy.RuntimeModel;
+
+namespace Microsoft.ReverseProxy.Middleware
+{
+    /// <summary>
+    /// Store current Cluster, ClusterConfig and Tracks proxy cluster destinations that are available to handle the current request.
```
Comment (Tratcher) [inline src/ReverseProxy/Middleware/IReverseProxyFeature.cs:11]:
```suggestion
    /// Store current ClusterConfig and Tracks proxy cluster destinations that are available to handle the current request.
```

File: src/ReverseProxy/Middleware/LoadBalancingMiddleware.cs [hunk 13]
```diff
@@ -31,23 +31,22 @@ internal class LoadBalancingMiddleware
 
         public Task Invoke(HttpContext context)
         { 
-            var cluster = context.GetRequiredCluster();
-            var destinationsFeature = context.GetRequiredDestinationFeature();
-            var destinations = destinationsFeature.Destinations;
+            var reverseProxy = context.GetReverseProxyFeature();
```
Comment (Tratcher) [inline src/ReverseProxy/Middleware/LoadBalancingMiddleware.cs:34]:
```suggestion
            var proxyFeature = context.GetReverseProxyFeature();
```
or reverseProxyFeature.

File: test/ReverseProxy.Tests/Middleware/LoadBalancerMiddlewareTests.cs [hunk 14]
```diff
@@ -71,19 +71,23 @@ public async Task Invoke_Works()
             httpContext.SetEndpoint(aspNetCoreEndpoint);
             Provide<ILoadBalancer, LoadBalancer>();
 
-            httpContext.Features.Set<IAvailableDestinationsFeature>(
-                new AvailableDestinationsFeature() { Destinations = new List<DestinationInfo>() { destination1, destination2 }.AsReadOnly() });
+            httpContext.Features.Set<IReverseProxyFeature>(
+                new ReverseProxyFeature()
+                {
+                    AvailableDestinations = new List<DestinationInfo>() { destination1, destination2 }.AsReadOnly(),
```
Comment (Tratcher) [inline test/ReverseProxy.Tests/Middleware/LoadBalancerMiddlewareTests.cs:77]: Arrays work too, right?
```suggestion
                    AvailableDestinations = new[] { destination1, destination2 },
```

File: src/ReverseProxy/Middleware/AffinitizedDestinationLookupMiddleware.cs [hunk 15]
```diff
@@ -35,30 +35,31 @@ internal class AffinitizedDestinationLookupMiddleware
...
+            var cluster = context.GetRequiredCluster();
+            return InvokeInternal(context, options, cluster.ClusterId);
         }
 
-        private async Task InvokeInternal(HttpContext context, ClusterConfig.ClusterSessionAffinityOptions options, ClusterInfo cluster)
+        private async Task InvokeInternal(HttpContext context, ClusterConfig.ClusterSessionAffinityOptions options, string clusterId)
         {
-            var destinationsFeature = context.GetRequiredDestinationFeature();
-            var destinations = destinationsFeature.Destinations;
+            var proxyFeature = context.GetRequiredProxyFeature();
```
Comment (Tratcher) [inline src/ReverseProxy/Middleware/AffinitizedDestinationLookupMiddleware.cs:53]: I didn't realize we were now getting the same proxy feature twice. This should be passed into InvokeInternal.

**Comment by Tratcher** [comment]
Thanks

**Pull request merged**